### ISSUE TITLE
Issues with switching between batch and non-batch predictions

### ISSUE THREAD

**Issue by Balandat** [cornellius-gp/gpytorch#340]
tl;dr: We need to better handle the precomputed cache for train-train covariance

Say you fit a GP in standard (non-batch) mode (`n x p` input).

- if you evaluate it in standard mode (`m x p` input), you're fine
- if you evaluate in batch mode (`b x m x p` input) **without having previously evaluated in standard mode**, you're fine
- if you first evaluate in standard mode, then in batch mode (or vice versa), an error is raised because of incompatible sizes of the train_train cache.

This is a bug, going back between the different modes should be possible.

Further, the following should be supported:
- fitting a GP in batch mode (`b x n x p` input) and then evaluating in "standard" mode (`b x m x p` input)
- fitting a GP in batch mode (`b x n x p` input)  and evaluating in batch-batch mode (`b' x b x n x p` input) 

The only thing that should not be supported is evaluating a GP fitted in batch mode (`b x n x p` input) on an incompatible size (`b' x n x p` input, where `b' != b`)

If the GP is fit in standard mode but evaluated in batched mode, we expand the train inputs here:
https://github.com/cornellius-gp/gpytorch/blob/2a96a881b77cb0f45d562e9146e89abbe4397591/gpytorch/models/exact_gp.py#L99

We then compute the same train-train covariance for each batch repeatedly during evaluation (b/c we use expand and Lazy tensors this is probably not terrible perf-wise, but unnecessary). Instead, what I would suggest is computing the train-train covariance cache once in standard mode, and then during the call of `exact_predictive_mean` do proper broadcasting here if necessary: https://github.com/cornellius-gp/gpytorch/blob/1a91ca523a24cfd9b6ad406c7994b0667b09177d/gpytorch/lazy/lazy_tensor.py#L483

This would likely require changes to how we deal with `full_mean` in the exact GP.

**Comment by jacobrgardner** [comment]
The actual bug parts of this issue are fixed via #342 -- the only remaining thing left to do is make it so that in the multitask setting we don't duplicate work if possible.

**Comment by Balandat** [comment]
So while this fixes forward passes, there are still issues with the backward pass, see https://github.com/cornellius-gp/gpytorch/pull/342#issuecomment-436725975
